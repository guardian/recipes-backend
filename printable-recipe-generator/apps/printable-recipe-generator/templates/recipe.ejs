<!DOCTYPE html>
<html>
<head>
    <meta charset="UTF-8"/>
    <title><%= title %></title>
    <style>
      body {
        font-family: Georgia, serif;
        font-size: 16px;
        line-height: 1.6;
        margin: 40px;
        color: #222;
      }

      .header {
        display: flex;
        justify-content: space-between;
        align-items: center;
      }

      .app-title {
        font-size: 24px;
        font-weight: bold;
      }

      .logo {
        height: 40px;
      }

      .spacer {
        margin: 2em 0;
      }

      .recipe-title {
        font-size: 22px;
        font-weight: bold;
      }

      .author {
        font-size: 16px;
        font-style: italic;
      }

      .labels {
        margin: 1em 0;
      }

      .label {
        display: inline-block;
        padding: 0.25em 0.6em;
        border-radius: 20px;
        font-size: 0.8em;
        text-transform: capitalize;
        color: white;
        font-weight: 500;
        margin-right: 6px;
      }

      .vegetarian {
        background-color: #228B22;
      }

      .GlutenFree {
        background-color: #FFA500;
        color: #3c2c00;
      }

      .dairy-free {
        background-color: #FFA500;
        color: #3c2c00;
      }

      .default {
        background-color: #999;
      }

      .timing, .serves {
        margin-bottom: 1em;
      }

      .timing p,
      .serves p {
        margin: 0.25em 0;
      }

      .section {
        margin-top: 3em;
      }

      .section-title {
        font-weight: bold;
        font-size: 1.1em;
        border-bottom: 2px solid #ccc;
        margin-bottom: 1em;
      }

      .ingredient-group {
        margin-bottom: 1em;
      }

      ul.instructions {
        list-style: disc inside;
        padding-left: 0;
      }

      .footer {
        margin-top: 4em;
        font-size: 14px;
        color: #555;
        text-align: center;
        border-top: 1px solid #ccc;
        padding-top: 1em;
      }
    </style>
</head>
<body>

<div class="header">
    <div class="app-title">Feast</div>
    <img src="guardianLogo.png" class="logo" alt="Company Logo"/>
</div>

<div class="spacer"></div>

<div class="recipe-title"><%= title %></div>
<% if (contributors) { %>
    <div class="author">By <%= contributors %></div>
<% } %>

<div class="spacer"></div>

<% if (suitableForDietIds && suitableForDietIds.length) { %>
    <div class="labels">
        <% for (const label of suitableForDietIds) {
            const className = label.replace(/\s/g, '').replace(/[^a-zA-Z]/g, '');
        %>
        <span class="label <%= className || 'default' %>"><%= label %></span>
        <% } %>
    </div>
<% } %>

<div class="timing">
    <% if (timings && timings.length) {
        const icons = {
            'prep-time': '⏱',
            'cook-time': '🔥',
            'marinate-time': '🧂'
        };
    %>
    <% for (const time of timings) { %>
        <p><strong><%= icons[time.qualifier] || '⏱' %> <%= time.text.split(' ')[0] %>
                :</strong> <%= time.text.split(' ').slice(1).join(' ') %></p>
    <% } %>
    <% } %>
</div>

<% if (serves && serves.length) { %>
    <div class="serves">
        <p><strong>¶
                Serves:</strong> <%= serves[0].amount.min === serves[0].amount.max ? serves[0].amount.min : `${serves[0].amount.min}-${serves[0].amount.max}` %> <%= serves[0].unit %>
        </p>
    </div>
<% } %>

<div class="spacer"></div>

<% if (description) { %>
    <div class="section">
        <div class="section-title">Description</div>
        <p><%= description %></p>
    </div>
<% } %>

<% if (ingredients && ingredients.length) { %>
    <div class="section">
        <div class="section-title">Ingredients</div>
        <% for (const group of ingredients) { %>
            <div class="ingredient-group">
                <ul>
                    <% for (const ing of group.ingredientsList) { %>
                        <li>
                            <%= ing.amount.min %>
                            <%= ing.unit || '' %>
                            <%= ing.name %>
                            <%= ing.suffix || '' %>
                            <% if (ing.optional) { %>(optional)
                            <% } %>
                        </li>
                    <% } %>
                </ul>
            </div>
        <% } %>
    </div>
<% } %>

<% if (instructions && instructions.length) { %>
    <div class="section">
        <div class="section-title">Method</div>
        <ul class="instructions">
            <% for (const step of instructions) { %>
                <li><%= step.description %></li>
            <% } %>
        </ul>
    </div>
<% } %>

</body>
</html>
