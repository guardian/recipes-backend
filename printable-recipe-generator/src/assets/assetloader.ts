import fs from 'fs';
import path from 'path';
import * as process from 'process';

const isTest =
	process.env.NODE_ENV === 'test' || process.env.JEST_WORKER_ID !== undefined;

const svgPath = (fileName: string) =>
	isTest
		? path.join(__dirname, 'svgs', fileName)
		: path.join(__dirname, 'src', 'assets', 'svgs', fileName);

export const svgs = {
	FeastLogo: fs.readFileSync(svgPath('FeastLogo.svg'), 'utf-8'),
	TheGuardianLogo: fs.readFileSync(svgPath('TheGuardianLogo.svg'), 'utf-8'),
	'dairy-free': fs.readFileSync(svgPath('Dairy-free.svg'), 'utf-8'),
	'gluten-free': fs.readFileSync(svgPath('Gluten-free.svg'), 'utf-8'),
	vegan: fs.readFileSync(svgPath('Vegan.svg'), 'utf-8'),
	vegetarian: fs.readFileSync(svgPath('Vegetarian.svg'), 'utf-8'),
	camera: fs.readFileSync(svgPath('camera.svg'), 'utf-8'),
	'feast-book-outlined': fs.readFileSync(
		svgPath('feast-book-outlined.svg'),
		'utf-8',
	),
	'clock-filled': fs.readFileSync(svgPath('clock-filled.svg'), 'utf-8'),
	'knife-and-fork': fs.readFileSync(svgPath('knife-and-fork.svg'), 'utf-8'),
};

//load fonts
const fontPath = (fileName: string) =>
	isTest
		? path.join(__dirname, 'fonts', fileName)
		: path.join(__dirname, 'src', 'assets', 'fonts', fileName);
export const fontsBase64 = {
	GuardianRegularEgyptianFont: fs
		.readFileSync(fontPath('GuardianTextEgyptian-Regular.ttf'))
		.toString('base64'),
	GuardianRegularSansFont: fs
		.readFileSync(fontPath('GuardianTextSans-Regular.ttf'))
		.toString('base64'),
	GuardianHeadlineSemiBoldFont: fs
		.readFileSync(fontPath('GHGuardianHeadline-Semibold.otf'))
		.toString('base64'),
	GuardianSansBoldFont: fs
		.readFileSync(fontPath('GuardianTextSans-Bold.ttf'))
		.toString('base64'),
};
