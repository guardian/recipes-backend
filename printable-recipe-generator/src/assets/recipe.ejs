<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8"/>
    <title><%= recipe.title %></title>
    <style>
      @font-face {
        font-family: 'GuardianRegularEgyptianFont';
        src: url("data:fonts/ttf;base64,<%= fontsBase64.RegularEgyptianFont %>") format('opentype');
      }
      @font-face {
        font-family: 'GuardianRegularSansFont';
        src: url("data:fonts/ttf;base64;<%= fontsBase64.RegularSansFont %>") format('opentype');
      }
      @font-face {
        font-family: 'GuardianHeadlineSemiBoldFont';
        src: url("data:fonts/otf;base64;<%= fontsBase64.HeadlineSemiBoldFont %>") format('truetype');
      }
      @font-face {
        font-family: 'GuardianSansBoldFont';
        src: url("data:fonts/ttf;base64;<%= fontsBase64.SansBoldFont %>") format('opentype');
      }

      body {
        font-family: 'GuardianRegularEgyptianFont', sans-serif;
        font-size: 17px;
        line-height: 1.6;
        margin: 40px;
        color: #20201D;
      }

      .header {
        font-family: 'GuardianHeadlineSemiBoldFont', sans-serif;
        font-size: 28px;
        display: flex;
        justify-content: space-between;
        align-items: center;
        margin-bottom: 40px;
      }

      .feast-logo {
        height: 40px;
      }

      .guardian-logo {
        height: 40px;
      }

      .recipe-title {
        font-size: 26px;
        font-weight: bold;
      }

      .author {
        font-family: 'GuardianRegularSansFont', sans-serif;
        font-size: 20px;
        margin-bottom: 20px;
        color: #3D4D26;
      }

      .labels {
        margin: 20px 0;
      }

      .timing-table {
        display: flex;
        flex-wrap: wrap;
        gap: 2rem;
        padding: 1rem 0;
        border-top: 1px solid #ccc;
        border-bottom: 1px solid #ccc;
        margin-bottom: 2rem;
      }

      .timing-row {
        font-family: 'GuardianRegularSansFont', sans-serif;
        font-size: 15px;
        display: flex;
        align-items: center;
        gap: 0.6rem;
      }

      .icon {
        width: 20px;
        height: 20px;
        vertical-align: middle;
      }

      .two-column-flow {
        column-count: 2;
        column-gap: 3rem;
        margin-top: 2rem;
      }

      .two-column-flow h3 {
        break-after: avoid;
        font-family: 'GuardianBoldSansFont', sans-serif;
        margin-top: 1rem;
      }

      .two-column-flow ul {
        list-style-type: disc;
        padding-left: 1.5rem;
        margin-bottom: 2rem;
      }

      .two-column-flow li {
        break-inside: avoid;
        font-family: 'GuardianRegularSansFont', sans-serif;
        font-size: 17px;
        margin-bottom: 0.5rem;
      }

      .recipe-footer {
        margin-top: 40px;
        padding-top: 20px;
        border-top: 1px solid #ddd;
        font-size: 14px;
        color: #444;
      }

      .footer-line {
        display: flex;
        align-items: flex-start;
        gap: 10px;
        margin-bottom: 10px;
      }

      .footer-icon svg {
        width: 18px;
        height: 18px;
        flex-shrink: 0;
        color: #666;
      }

      .footer-text a {
        font-family: 'GuardianRegularSansFont', sans-serif;
        font-size: 15px;
        text-decoration: none;
      }

      h3 {
        font-weight: bold;
      }
    </style>
</head>
<body>

<div class="header">
    <div class="feast-logo">
        <%- svgs.FeastLogo %>
    </div>
    <div class="guardian-logo">
        <%- svgs.TheGuardianLogo %>
    </div>
</div>

<div class="recipe-title"><%= recipe.title %></div>
<% if (recipe.contributors && typeof chefs === 'object' && chefs?.[recipe.contributors]) { %>
<div class="author">By <%=  chefs[recipe.contributors].webTitle %></div>
<% } %>
<% if (recipe.contributors && typeof chefs != 'object' ) { %>
    <div class="author">By <%=  recipe.contributors %></div>
<% } %>


<% if (recipe.suitableForDietIds && recipe.suitableForDietIds.length) { %>
    <div class="labels">
        <% for (const label of recipe.suitableForDietIds) {
            const labelKey = label.toLowerCase().replace(/\s+/g, '-');
        %>
        <span>
            <%- svgs[labelKey] %>
      </span>
        <% } %>
    </div>
<% } %>

<div class="timing-table">
    <% for (const time of recipe.timings || []) { %>
        <div class="timing-row">
            <div>
                <% if (time.qualifier === 'prep-time') { %>
                    <svg class="icon" viewBox="0 0 24 24">
                        <circle cx="12" cy="12" r="10" stroke="#000" stroke-width="2" fill="none"/>
                        <path d="M12 6v6l4 2" stroke="#000" stroke-width="2" fill="none"/>
                    </svg>
                <% } else if (time.qualifier === 'cook-time') { %>
                    <svg class="icon" viewBox="0 0 24 24">
                        <path d="M4 10h16v2H4z" fill="#000"/>
                        <circle cx="6" cy="16" r="2" fill="#000"/>
                        <circle cx="18" cy="16" r="2" fill="#000"/>
                    </svg>
                <% } else if (time.qualifier === 'marinate-time') { %>
                    <svg class="icon" viewBox="0 0 24 24">
                        <path d="M12 3v18M5 12h14" stroke="#000" stroke-width="2" fill="none"/>
                    </svg>
                <% } %>
            </div>
            <div><%= time.text %></div>
        </div>
    <% } %>

    <% if (recipe.serves && recipe.serves.length) { %>
        <div class="timing-row">
            <svg class="icon" viewBox="0 0 24 24">
                <path d="M4 4h16v2H4zM4 8h16v2H4zM4 12h16v2H4z" fill="#000"/>
            </svg>
            <div>Serves <%= recipe.serves[0].amount.min %> <%= recipe.serves[0].unit %></div>
        </div>
    <% } %>
</div>

<div class="description"><%= recipe.description %></div>

<div class="two-column-flow">
    <div class="column ingredients">
        <h3>Ingredients</h3>
        <ul>
            <% for (const group of recipe.ingredients || []) { %>
                <% for (const ing of group.ingredientsList || []) { %>
                    <li>
                        <%= ing.amount?.min %>
                        <% if (ing.unit) { %> <%= ing.unit %>
                        <% } %>
                        <%= ing.name %>
                        <% if (ing.suffix) { %> - <%= ing.suffix %>
                        <% } %>
                    </li>
                <% } %>
            <% } %>
        </ul>
    </div>

    <div class="column method">
        <h3>Method</h3>
        <ul>
            <% for (const step of recipe.instructions || []) { %>
                <li><%= step.description %></li>
            <% } %>
        </ul>
    </div>
</div>

<% if (recipe.bookCredit || recipe.featuredImage) { %>
<footer class="recipe-footer">
    <div class="footer-line">
        <% if (recipe.bookCredit) { %>
    <span class="footer-icon">
      <%- svgs["feast-book-outlined"] %>
    </span>
        <span class="footer-text">
        <%= recipe.bookCredit %>>
    </span>
        <% } %>
    </div>
    <div class="footer-line">
        <% if (recipe.featuredImage) { %>
    <span class="footer-icon">
      <%- svgs.camera %>
    </span>

        <span class="footer-text">
        <%= recipe.featuredImage.source %>>
    </span>
        <% } %>
    </div>
</footer>
<% } %>


</body>
</html>
