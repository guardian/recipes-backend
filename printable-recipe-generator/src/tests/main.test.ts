import fs from 'fs';
import path from 'path';
import * as ejs from 'ejs';
import { fontsBase64, svgs } from '../assets/assetloader';
import recipe from '../data/sampleRecipe.json';

//load Contributors
interface ChefData {
	webTitle: string;
	webUrl: string;
	apiUrl: string;
	bio?: string;
	bylineImageUrl?: string;
	bylineLargeImageUrl?: string;
}

const chefs: Record<string, ChefData> = {
	'profile/yotamottolenghi': {
		webTitle: 'Yotam Ottolenghi',
		webUrl: 'http://www.code.dev-theguardian.com/profile/yotamottolenghi',
		apiUrl: 'http://content.code.dev-guardianapis.com/profile/yotamottolenghi',
		bio: "<p>Originally from Jerusalem, Yotam Ottolenghi is a London-based restaurateur and food writer. He runs Nopi and Ottolenghi, and is the author of three bestselling books; his fourth book, <a href='http://www.guardianbookshop.co.uk/BerteShopWeb/viewProduct.do?ISBN=9780091957155\" title=''>Plenty More</a>, is published in September 2014</p>",
		bylineImageUrl:
			'https://static.guim.co.uk/sys-images/Guardian/Pix/pictures/2014/4/17/1397749341692/YotamOttolenghi.jpg',
		bylineLargeImageUrl:
			'https://static.guim.co.uk/sys-images/Guardian/Pix/pictures/2014/3/13/1394733751371/YotamOttolenghi.png',
	},
	'profile/nigelslater': {
		webTitle: 'Nigel Slater',
		webUrl: 'http://www.code.dev-theguardian.com/profile/nigelslater',
		apiUrl: 'http://content.code.dev-guardianapis.com/profile/nigelslater',
		bio: "<p>Nigel Slater has been the Observer's food writer for 20 years. His cookery books, which include Appetite, Eat and the Kitchen Diaries have won a host of awards, while his autobiography Toast â€“ The Story of a Boy's Hunger was adapted by BBC Films, starring Helena Bonham Carter and Freddie Highmore</p>",
		bylineImageUrl:
			'https://static.guim.co.uk/sys-images/Guardian/Pix/pictures/2014/4/17/1397749337461/NigelSlaterLv2.jpg',
		bylineLargeImageUrl:
			'https://static.guim.co.uk/sys-images/Guardian/Pix/pictures/2014/4/17/1397762419839/NigelSlater.png',
	},
	'profile/claireptak': {
		webTitle: 'Claire Ptak',
		webUrl: 'http://www.code.dev-theguardian.com/profile/claireptak',
		apiUrl: 'http://content.code.dev-guardianapis.com/profile/claireptak',
	},
};
describe('Sample recipe ', () => {
	it('should match snapshot', () => {
		const templatePath = path.join(__dirname, '../assets/recipe.ejs');
		const template = fs.readFileSync(templatePath, 'utf-8');
		const html = ejs.render(template, {
			recipe,
			svgs,
			fontsBase64,
			chefs,
			qrImageDataUrl: '',
		});
		expect(html).toMatchSnapshot();
	});

	it('should match snapshot even when chefs are undefined', () => {
		const templatePath = path.join(__dirname, '../assets/recipe.ejs');
		const template = fs.readFileSync(templatePath, 'utf-8');
		const html = ejs.render(template, {
			recipe,
			svgs,
			fontsBase64,
			chefs: undefined,
			qrImageDataUrl: '',
		});
		expect(html).toMatchSnapshot();
	});

	it('should match snapshot even when chefs list is null', () => {
		const templatePath = path.join(__dirname, '../assets/recipe.ejs');
		const template = fs.readFileSync(templatePath, 'utf-8');
		const html = ejs.render(template, {
			recipe,
			svgs,
			fontsBase64,
			chefs: null,
			qrImageDataUrl: '',
		});
		expect(html).toMatchSnapshot();
	});

	it('should match snapshot even when chefs list is empty', () => {
		const templatePath = path.join(__dirname, '../assets/recipe.ejs');
		const template = fs.readFileSync(templatePath, 'utf-8');
		const html = ejs.render(template, {
			recipe,
			svgs,
			fontsBase64,
			chefs: [],
			qrImageDataUrl: '',
		});
		expect(html).toMatchSnapshot();
	});

	it('should match snapshot even when chefs list contains more then 1 chef', () => {
		const templatePath = path.join(__dirname, '../assets/recipe.ejs');
		const template = fs.readFileSync(templatePath, 'utf-8');
		const html = ejs.render(template, {
			recipe,
			svgs,
			fontsBase64,
			chefs: chefs,
			qrImageDataUrl: '',
		});
		expect(html).toContain('Yotam Ottolenghi');
		expect(html).toContain('Nigel Slater');
		expect(html).toMatchSnapshot();
	});

	it('should match snapshot when QR code is present', () => {
		const templatePath = path.join(__dirname, '../assets/recipe.ejs');
		const template = fs.readFileSync(templatePath, 'utf-8');
		const qrImageDataUrl =
			'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAKQAAACkCAYAAAAZtYVBAAAAAklEQVR4AewaftIAAAYsSURBVO3BQY4kRxLAQDLQ//8yd45+SqCQ1aOQ1s3sD9a6xGGtixzWushhrYsc1rrIYa2LHNa6yGGtixzWushhrYsc1rrIYa2LHNa6yGGtixzWushhrYv88JLK31TxhspUMalMFU9UpoonKlPFpPKJiicqf1PFG4e1LnJY6yKHtS7yw5dVfJPKJ1SmiqniEyqfUJkqpopPVHxTxTepfNNhrYsc1rrIYa2L/PDLVD5R8QmVqWJSmSomlScVk8pU8URlqnhS8URlqnhD5RMVv+mw1kUOa13ksNZFfviPq5hUpoonKk9UPqEyVUwqn1CZKv7NDmtd5LDWRQ5rXeSH/xiVb6qYVKaKJypTxScq/p8c1rrIYa2LHNa6yA+/rOJvqniiMqk8qXii8obKVPEJlaniExU3Oax1kcNaFzmsdZEfvkzlJipTxaQyVUwqU8WkMlVMKm+oTBVvqNzssNZFDmtd5LDWRX54qeLfRGWq+ITKVPGkYlKZKiaVqeKNin+Tw1oXOax1kcNaF/nhJZWpYlL5poqpYlJ5UjGpvKHyiYpJZaqYVKaKT6h8U8VvOqx1kcNaFzmsdZEfXqqYVKaKT6hMFZ+omFSeVPxNKlPFpDJVTCpvVDxReaLypOKNw1oXOax1kcNaF7E/+CKVT1Q8UXlS8UTlExVPVKaKSeVJxaQyVUwqU8UbKm9U/KbDWhc5rHWRw1oX+eGXVbxRMalMKk8qJpWp4onKE5Wp4onKE5WpYlKZKt6omFQ+oTJVvHFY6yKHtS5yWOsi9gcvqDypmFTeqHiiMlV8k8pUMalMFZPKk4onKk8q/s0Oa13ksNZFDmtd5Icvq5hUpoonKlPFE5WpYlL5RMWk8kTlicpU8UTlExVPVJ5UPFGZKiaVqeKNw1oXOax1kcNaF/nhH6byROUTKr+p4hMq/6SKNyomlanimw5rXeSw1kUOa13kh19WMalMFZPKVPFEZaqYVH6TylTxROWbVJ5UTCpTxaTypGJSmSreOKx1kcNaFzmsdZEfXqqYVJ5UfEJlqniiMlVMKr9JZap4Q2WqmFQ+UTGpTBVPVKaKbzqsdZHDWhc5rHWRHy5XMal8QmWqmFTeqHii8qRiUpkqJpVPqEwVU8UTlScqU8Ubh7UucljrIoe1LmJ/8ILKk4o3VJ5UTCpvVDxRmSomlaniicpUMalMFZPKVDGpfFPFbzqsdZHDWhc5rHWRH16qmFTeUJkqJpVPVEwq36QyVUwqU8UTlaliUvmmiknln3RY6yKHtS5yWOsiP3xZxaQyVUwqU8WkMlW8UTGpPFF5UvGk4g2VJxWTylQxqUwqTyomlanimw5rXeSw1kUOa13E/uAFlaniicqTiicqTyq+SeVJxTepPKn4hMpU8QmVqWJSmSreOKx1kcNaFzmsdZEffpnKVPEJlU+oPKn4RMWk8kTlScWTik+oTBVvqEwVk8pU8U2HtS5yWOsih7UuYn/wgspU8UTlExWfUPmbKj6h8kbFJ1SeVDxRmSomlanijcNaFzmsdZHDWhexP3hBZar4hMpUMan8popPqHyi4g2Vf1LFpDJVfNNhrYsc1rrIYa2L2B/8i6k8qXhDZar4hMobFZPKk4pPqEwVk8onKt44rHWRw1oXOax1kR9eUvmbKqaKN1SmiqliUpkqJpUnFZ9QeUNlqniiMlX8TYe1LnJY6yKHtS7yw5dVfJPKE5WpYlKZKj6h8kTlScUnVKaKJypPKt5QeVLxTYe1LnJY6yKHtS7ywy9T+UTFP0nljYpJZaqYVL5J5ZsqJpXfdFjrIoe1LnJY6yI//MdVTCpPKiaVqWJSeVLxpOKJyhsVk8onVKaKSWWqeOOw1kUOa13ksNZFfviPUZkq3qiYVJ6ofFPFpPKkYlKZKp6oTBVPKr7psNZFDmtd5LDWRewPXlCZKr5JZar4N1OZKiaVJxWfUHlS8URlqvhNh7UucljrIoe1LmJ/8ILK31QxqUwVT1SeVDxRmSomlaniicpUMak8qfgmlaliUpkqvumw1kUOa13ksNZF7A/WusRhrYsc1rrIYa2LHNa6yGGtixzWushhrYsc1rrIYa2LHNa6yGGtixzWushhrYsc1rrIYa2L/A8vMQB13c2jwAAAAABJRU5ErkJggg==';
		const html = ejs.render(template, {
			recipe,
			svgs,
			fontsBase64,
			chefs,
			qrImageDataUrl,
		});
		expect(html).toMatchSnapshot();
	});

	it('should match snapshot when QR code is not present', () => {
		const templatePath = path.join(__dirname, '../assets/recipe.ejs');
		const template = fs.readFileSync(templatePath, 'utf-8');
		const html = ejs.render(template, {
			recipe,
			svgs,
			fontsBase64,
			chefs,
			qrImageDataUrl: null,
		});
		expect(html).toMatchSnapshot();
	});
});
