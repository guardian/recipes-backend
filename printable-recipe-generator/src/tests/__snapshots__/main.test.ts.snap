// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sample recipe  should match snapshot 1`] = `
"<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8"/>
    <title>Mixed vegetable and potato fritters with harissa</title>
    <style>
      body {
        font-family: 'Georgia', serif;
        font-size: 16px;
        line-height: 1.6;
        margin: 40px;
        color: #222;
      }

      .header {
        display: flex;
        justify-content: space-between;
        align-items: center;
        margin-bottom: 40px;
      }

      .app-title {
        font-size: 24px;
        font-weight: bold;
      }

      .logo {
        height: 40px;
      }

      .recipe-title {
        font-size: 26px;
        font-weight: bold;
      }

      .author {
        font-size: 16px;
        font-style: italic;
        margin-bottom: 20px;
      }

      .labels {
        margin: 20px 0;
      }

      .label {
        display: inline-block;
        padding: 4px 10px;
        margin-right: 6px;
        border-radius: 20px;
        font-size: 14px;
        font-weight: 600;
        color: white;
      }

      .label-vegetarian {
        background-color: #2f6348;
      }

      .label-dairy-free {
        background-color: #f5c344;
        color: #3c2c00;
      }

      .timing-table {
        display: flex;
        flex-wrap: wrap;
        gap: 2rem;
        padding: 1rem 0;
        border-top: 1px solid #ccc;
        border-bottom: 1px solid #ccc;
        margin-bottom: 2rem;
      }

      .timing-row {
        display: flex;
        align-items: center;
        gap: 0.6rem;
        font-size: 14px;
      }

      .icon {
        width: 20px;
        height: 20px;
        vertical-align: middle;
      }

      .two-columns {
        display: flex;
        justify-content: space-between;
        gap: 3rem;
      }

      .column {
        width: 50%;
      }

      .ingredients ul,
      .steps ul {
        padding-left: 1.5rem;
      }

      .ingredients li,
      .steps li {
        margin-bottom: 0.5rem;
      }

      .footer {
        border-top: 1px solid #ccc;
        margin-top: 3rem;
        padding-top: 1rem;
        font-size: 14px;
        color: #777;
      }

      .footer div {
        margin-bottom: 0.3rem;
      }
    </style>
</head>
<body>

<div class="header">
    <div class="app-title">Feast</div>
    <img src="guardianLogo.png" class="logo" alt="Company Logo"/>
</div>

<div class="recipe-title">Mixed vegetable and potato fritters with harissa</div>

    <div class="author">By profile/yotamottolenghi</div>



    <div class="labels">
        
            <span class="label label-vegetarian">vegetarian</span>
        
            <span class="label label-dairy-free">dairy-free</span>
        
    </div>


<div class="timing-table">
    
        <div class="timing-row">
            <div>
                
                    <svg class="icon" viewBox="0 0 24 24">
                        <circle cx="12" cy="12" r="10" stroke="#000" stroke-width="2" fill="none"/>
                        <path d="M12 6v6l4 2" stroke="#000" stroke-width="2" fill="none"/>
                    </svg>
                
            </div>
            <div>Prep 25 min</div>
        </div>
    
        <div class="timing-row">
            <div>
                
                    <svg class="icon" viewBox="0 0 24 24">
                        <path d="M4 10h16v2H4z" fill="#000"/>
                        <circle cx="6" cy="16" r="2" fill="#000"/>
                        <circle cx="18" cy="16" r="2" fill="#000"/>
                    </svg>
                
            </div>
            <div>Cook 1 hr 20 min</div>
        </div>
    

    
        <div class="timing-row">
            <svg class="icon" viewBox="0 0 24 24">
                <path d="M4 4h16v2H4zM4 8h16v2H4zM4 12h16v2H4z" fill="#000"/>
            </svg>
            <div>Serves 16 fritters</div>
        </div>
    
</div>

<div class="description">These are inspired by Moroccan maakouda, a fried potato street food snack. It’s usually stuffed into a sandwich with lots of harissa, although you can also serve it with yoghurt, to make it more kid-friendly. Serve warm or at room temperature: they won’t be as crisp once they’ve sat around for a bit, but they’ll still taste great</div>

<div class="two-columns">
    <div class="column ingredients">
        <h3>Ingredients</h3>
        <ul>
            
                
                    <li>
                        2
                        
                        baking potatoes,
                         - peeled and cut into 3cm chunks (500g net weight)
                        
                    </li>
                
                    <li>
                        
                        
                        Salt and black pepper
                        
                    </li>
                
                    <li>
                        250
                         g
                        
                        frozen mixed vegetables,
                         - steamed according to packet directions
                        
                    </li>
                
                    <li>
                        1
                         clove
                        
                        garlic,
                         - roughly chopped
                        
                    </li>
                
                    <li>
                        10
                         tbsp
                        
                        flat-leaf parsley,
                         - roughly chopped
                        
                    </li>
                
                    <li>
                        2
                         tsp
                        
                        cumin seeds,
                         - toasted and roughly crushed in a mortar
                        
                    </li>
                
                    <li>
                        1
                        
                        egg
                        
                    </li>
                
                    <li>
                        2
                         tsp
                        
                        lemon zest
                        
                    </li>
                
                    <li>
                        30
                         g
                        
                        plain flour,
                         - for coating
                        
                    </li>
                
                    <li>
                        700
                         ml
                        
                        sunflower oil,
                         - for deep frying
                        
                    </li>
                
            
                
                    <li>
                        1.5
                         tbsp
                        
                        rose harissa
                        
                    </li>
                
                    <li>
                        2
                         tbsp
                        
                        lemon juice
                        
                    </li>
                
                    <li>
                        1
                         tsp
                        
                        runny honey
                        
                    </li>
                
                    <li>
                        2
                        
                        garlic cloves
                         - peeled and crushed
                        
                    </li>
                
                    <li>
                        2
                         tbsp
                        
                        olive oil
                        
                    </li>
                
            
                
                    <li>
                        85
                         g
                        
                        plain flour
                        
                    </li>
                
                    <li>
                        40
                         g
                        
                        corn flour
                        
                    </li>
                
                    <li>
                        1.25
                         tsp
                        
                        instant dried yeast,
                         - or 12g fresh yeast, roughly crumbled
                        
                    </li>
                
                    <li>
                        0.66
                         tsp
                        
                        baking powder
                        
                    </li>
                
                    <li>
                        0.66
                         tsp
                        
                        ground turmeric
                        
                    </li>
                
                    <li>
                        135
                         ml
                        
                        lukewarm water
                        
                    </li>
                
            
        </ul>
    </div>

    <div class="column steps">
        <h3>Method</h3>
        <ul>
            
                <li>Put the potatoes and two teaspoons of salt in a medium saucepan for which you have a lid, add enough water to cover by about 3cm, then bring to a boil on a medium-high heat. Lower the heat to medium, cover and leave to cook for eight minutes, or until the potatoes are easily pierced with a knife. Drain well, then, once they’re cool enough to handle, use a potato ricer to rice the potatoes into a large bowl (if you don’t have a ricer, use a masher).</li>
            
                <li>While the potatoes are cooking, put the mixed vegetables, garlic and parsley in a food processor and pulse a few times, until finely chopped but not pureed. Add this to the mashed potato bowl, then mix in the cumin, egg, lemon zest, a teaspoon of salt and a good grind of pepper and refrigerate for 30 minutes, so the mix firms up slightly.</li>
            
                <li>Whisk all the dip ingredients and a pinch of salt in a small bowl.</li>
            
                <li>Make the batter by whisking all the ingredients in a medium bowl, then leave to bubble up for about 20 minutes.</li>
            
                <li>Meanwhile, put the 30g flour in a shallow bowl. Use your hands to make roughly 5cm-diameter, 40-45g patties out of the potato mixture – you should end up with about 16 patties, and don’t worry if they’re not perfect. Dip each patty into the flour to coat, shaking off the excess, and put on an oven tray.</li>
            
                <li>Heat the sunflower oil in a medium-sized high-sided saute pan over a medium-high heat. Once the oil is hot, and working one patty at a time, dip a patty into the batter to coat, then drop into the hot oil (don’t worry if there are scraggly bits of batter – this just adds to the look). Fry the patties in three batches, for about six or seven minutes each, until deeply golden. Use a slotted spoon to transfer to a tray lined with kitchen paper and repeat with the remaining patties and batter.</li>
            
                <li>Transfer the fritters to a large plate and serve hot with the harissa dip in a bowl alongside.</li>
            
        </ul>
    </div>
</div>

</body>
</html>
"
`;
